autopropose.py

```diff
@@ -1,71 +1,138 @@
         key, value = line.split('=', maxsplit=1)
-        value = value.strip('"')
-        value = value.strip("'")
+        value = value.strip('"').strip("'")
         result[key] = value
     return result
 
 
-def parse_conf_file(file_name):
+def parse_conf_file(file_name: str) -> Dict[str, str]:
     with open(file_name) as file:
         contents = file.read()
         return parse_conf(contents)
 
 
-def main():
-    LOGGER.setLevel(logging.INFO)
+def deploy(contract: str, private_key: str) -> None:
+    deploy_command = [
+        'docker', 'exec', 'rnode',
+        './bin/rnode',
+        '-c', '/var/lib/rnode-static/rnode.conf',
+        'deploy',
+        '--phlo-limit', '1000000000',
+        '--phlo-price', '1',
+        '--private-key', private_key,
+        contract,
+    ]
+
+    try:
+        LOGGER.info("Deploying %s", contract)
+        subprocess.check_output(deploy_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
+        LOGGER.exception('deploy')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
+
+
+
+def propose() -> None:
+    propose_command = [
+        'docker',
+        'exec',
+        'rnode',
+        './bin/rnode',
+        '-c', '/var/lib/rnode-static/rnode.conf',
+        'propose',
+    ]
+
+    try:
+        LOGGER.info("Proposing...")
+        subprocess.check_output(propose_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
+        LOGGER.exception('propose')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
+
+
+@contextlib.contextmanager
+def must_take_at_least(seconds: int) -> Iterator[None]:
+    started_at = int(time.monotonic())
+    try:
+        yield
+    finally:
+        finished_at = int(time.monotonic())
+        took = finished_at - started_at
+        if took < seconds:
+            wait_duration = seconds - took
+            time.sleep(wait_duration)
 
-    config = parse_conf_file('/var/lib/rnode-static/autopropose.conf')
-    deploy_no_sooner_than_every = int(config.get('period', '60'))
 
-    deploy_env = {}
-    if 'contract' in config:
-        deploy_env['contract'] = config['contract']
-    elif 'RD_OPTION_CONTRACT' in os.environ:
-        deploy_env['RD_OPTION_CONTRACT'] = os.environ['RD_OPTION_CONTRACT']
 
-    for env_var in ('RD_OPTION_DEPLOY_KEY', 'RD_OPTION_RNODE_LAUNCHER_ARGS', 'RD_OPTION_RNODE_DEPLOY_ARGS'):
-        if env_var in os.environ:
-            deploy_env[env_var] = os.environ[env_var]
+def main() -> int:
+    config = parse_conf_file('/var/lib/rnode-static/autopropose.conf')
 
-    LOGGER.info('Deploying contract %s every %d seconds', deploy_env.get('contract'), deploy_no_sooner_than_every)
+    contract = config.get('contract')
+    if contract is None or contract == '' or contract == 'none':
+        LOGGER.info('No contract configured.  Going to sleep...')
+        while True:
```

Comment:
I don't get this. If no contract is configured, then `main()` effectively loops infinitely? It should still propose every `period`.